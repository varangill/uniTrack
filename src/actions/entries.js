import uuid from 'uuid';
import database from '../firebase/firebase.js';

export const initAddEntry = (entryData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      description = '', //these are default values for if nothing is passed in
      note = '',
      tag = "task",
      createdAt = 0
    } = entryData;
    const entryToAdd = {description, note, tag, createdAt}

    return database.ref(`users/${uid}/entries`).push(entryToAdd).then((ref) => { 
      dispatch(addEntry({
        id: ref.key,
        ...entryToAdd
      }));
    });
  };
};

// ADD_ENTRY
export const addEntry = (entry) => ({ 
  type: 'ADD_ENTRY',
  entry
});

// REMOVE_ENTRY
export const removeEntry = ({ id } = {}) => ({
  type: 'REMOVE_ENTRY',
  id
});

export const initRemoveEntry = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/entries/${id}`).remove().then(() => {
      dispatch(removeEntry({id}));
    });
  };
};

// EDIT_ENTRY
export const editEntry = (id, updates) => ({
  type: 'EDIT_ENTRY',
  id,
  updates
});

export const initEditEntry = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/entries/${id}`).update(updates).then(() => {
      dispatch(editEntry(id, updates));
    })
  }
};

// SET_ENTRIES
export const setEntries = (entries) => ({
  type: 'SET_ENTRIES',
  entries
})

export const startSetEntries = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/entries`).once('value').then((snapshot) => {
      const entries = [];

      snapshot.forEach((childSnapshot) => {
        entries.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });

      dispatch(setEntries(entries));
    });
  };
};