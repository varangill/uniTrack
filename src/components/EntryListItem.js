import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';

const EntryListItem = ({ id, description, tag, createdAt }) => (
  <Link className="list-entry" to={`/edit/${id}`}>
    <div>
      <h3 className="list-entry__desc">{description}</h3>
      <span className="list-item__date">{moment(createdAt).format('MMM Do, YYYY')}</span>
    </div>
    <h4 className="list-item__tag">{tag}</h4>
  </Link>
);

export default EntryListItem;

// ATTEMPTS AT PIN FUNCTION
// export default class EntryListItem extends React.Component {
//   constructor(props) {
//     super(props);
//     this.state = {
//       description: props.entry ? props.entry.description : '',
//       note: props.entry ? props.entry.note : '',
//       alphabet: props.entry ? (props.entry.alphabet / 100).toString() : '',
//       tag: props.entry ? props.entry.tag : 'task',
//       createdAt: props.entry ? moment(props.entry.createdAt) : moment()
//     };
//   }
//   pinFunc = () => {
//     console.log("pinned!");
//   };
//   render () {
//     return (
//       <div>
//         <Link to={`/edit/${this.state.id}`}>
//           <h3>{this.state.description}</h3>
//         </Link>
//         <p>{this.state.alphabet} - {this.state.createdAt} - {this.state.tag}</p>
//       </div>
//     )
//   }
// }
