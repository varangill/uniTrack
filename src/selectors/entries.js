import moment from 'moment';

// Get visible entries

export default (entries, { text, sortBy, startDate, endDate, tag }) => {
  return entries.filter((entry) => {
    const createdAtMoment = moment(entry.createdAt);
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
    const textMatch = entry.description.toLowerCase().includes(text.toLowerCase());
    const tagMatch = tag === 'all' ? true : tag === entry.tag;
    
    return startDateMatch && endDateMatch && textMatch && tagMatch;
  }).sort((a, b) => {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if (sortBy === 'alphabet') { //this is alphabet now
      return a.description.localeCompare(b.description);
    }
  });
};
