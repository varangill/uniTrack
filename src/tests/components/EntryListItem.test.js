import React from 'react';
import { shallow } from 'enzyme';
import entries from '../fixtures/entries';
import EntryListItem from '../../components/EntryListItem';

test('should render EntryListItem correctly', () => {
  const wrapper = shallow(<EntryListItem {...entries[0]} />);
  expect(wrapper).toMatchSnapshot();
});
