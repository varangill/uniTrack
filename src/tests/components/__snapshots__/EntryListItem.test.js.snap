// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EntryListItem correctly 1`] = `
<Link
  className="list-entry"
  replace={false}
  to="/edit/1"
>
  <div>
    <h3
      className="list-entry__desc"
    >
      Gum
    </h3>
    <span
      className="list-item__date"
    >
      Dec 31st, 1969
    </span>
  </div>
  <h4
    className="list-item__tag"
  >
    task
  </h4>
</Link>
`;
